makeWorkspace/make_ws.py

```diff
@@ -31,6 +31,85 @@ def cli_args():
 
     return args
 
+def create_workspace(fin, f_jes, fout, category):
+  '''Create workspace and write the relevant histograms in it for the given category, returns the workspace.'''
+  fdir = fin.Get("category_"+category)
+  foutdir = fout.mkdir("category_"+category)
+
+  wsin_combine = ROOT.RooWorkspace("wspace_"+category,"wspace_"+category)
+  wsin_combine._import = SafeWorkspaceImporter(wsin_combine)
+
+  variable_name = "mjj" if ("vbf" in category) else "met"
+  varl = ROOT.RooRealVar(variable_name, variable_name, 0,100000);
+  # Helper function
+  def write_obj(obj, name):
+    '''Converts histogram to RooDataHist and writes to workspace + ROOT file'''
+    print "Creating Data Hist for ", name
+    dhist = ROOT.RooDataHist(
+                          name,
+                          "DataSet - %s, %s"%(category,name),
+                          ROOT.RooArgList(varl),
+                          obj
+                          )
+    wsin_combine._import(dhist)
+
+    # Write the individual histograms
+    # for easy transfer factor calculation
+    # later on
+    obj.SetDirectory(0)
+    foutdir.cd()
+    foutdir.WriteTObject(obj)
+
+  # Loop over all keys in the input file
+  # pick out the histograms, and add to
+  # the work space.
+  keys_local = fdir.GetListOfKeys()
+  for key in keys_local:
+    obj = key.ReadObj()
+    if type(obj) not in [ROOT.TH1D, ROOT.TH1F]:
+      continue
+    title = obj.GetTitle()
+    name = obj.GetName()
+
+    # Ensure non-zero integral for combine
+    if not obj.Integral() > 0:
+      obj.SetBinContent(1,0.0001)
+
+    # Add overflow to last bin
+    overflow        = obj.GetBinContent(obj.GetNbinsX()+1)
+    overflow_err    = obj.GetBinError(obj.GetNbinsX()+1)
+    lastbin         = obj.GetBinContent(obj.GetNbinsX())
+    lastbin_err     = obj.GetBinError(obj.GetNbinsX())
+    new_lastbin     = overflow + lastbin
+    new_lastbin_err = sqrt(overflow_err**2 + lastbin_err**2)
+
+    obj.SetBinContent(obj.GetNbinsX(), new_lastbin)
+    obj.SetBinError(obj.GetNbinsX(), new_lastbin_err)
+    obj.SetBinContent(obj.GetNbinsX()+1, 0)
+    obj.SetBinError(obj.GetNbinsX()+1, 0)
+
+    write_obj(obj, name)
+    
+    # Systematic variations
+    if not 'data' in name:
+      channel = re.sub("(loose|tight)","", category)
```

Comment:
Should / can we refactor everything inside the if block here into a separate function? ```add_jes_variations(...)```?

Reply:
I'm thinking this because I have a PR lined up that would add diboson uncertainty variations. Will be cleaner if we can just have:
```
add_jes_variations(...)
add_diboson_variations(...)
...
```

Reply:
Sure sounds good, I can re-factor the code to a new function to get JES variations